app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,26 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def google_oauth2
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    if @user.persisted?
+      flash[:notice] = "devise.omniauth_callbacks.success", :kind => "Google"
+      sign_in_and_redirect @user, :event => :authentication
+    else
```

Comment:
I believe this second scenario basically amounts to some sort of failure during #from_omniauth, so it doesn't even seem necessary if we go with #first_or_create!, since an exception thrown there would prevent us from getting to this point in the code. We could leave it just in case, but I would add an appropriate alert to the flash and not bother with saving the response to the session unless we want to somehow try and use it later on.